### ISSUE TITLE
PUT: Support optimistic locking using E-Tag in the If-Match header

### ISSUE THREAD

**Issue by guruz** [owncloud/core#4511]
See this client issue: https://github.com/owncloud/mirall/issues/881

Basically, we want to avoid that a client overwrites a file while there had been a change to it in the meantime.

**Comment by PVince81** [comment]
@DeepDiver1975 is this still relevant in the age of transactional file locking and where the lock API of Webdav has been removed ?

**Comment by DeepDiver1975** [comment]
Id say: yes - close

@guruz

**Comment by guruz** [comment]
So you are handling ```If-Match```?

**Comment by ogoffart** [comment]
Are you sure that two clients sending the same file with the same with the same If-Match header, only one will succeed, and the other will error out with 412?   If that's the case, then this issue can be closed.

**Comment by PVince81** [comment]
The operation isn't atomic unfortunately. If the If-Match header check was done in parallel for two highly parallel clients and both passed, they will both continue uploading at the same time until the final rename.

If say file 1 is bigger and file 2 is smaller, file 2 would reach the final rename without any locking impairment. So the final file becomes file 2. But wait, file 1 is still transfering and would also reach the final rename and then overwrite.

To fix this, we'd need to redo the "if-match" check again before the final rename.
Or at least check what the old etag was before doing the final rename. Then, inside of the lock, see if the old etag changed before overwriting.

tldr; still an issue
